File: examples/javascript/convert.js
/**
 * MegaConvert API — Node.js example using fetch
 * Full docs: https://megaconvert.io/docs/api
 *
 * Requires Node.js 18+ (native fetch)
 */

const fs = require("fs");
const path = require("path");

const API_KEY = "mc_your_api_key_here";
const BASE_URL = "https://megaconvert.io/api/v1";

const headers = { "X-API-Key": API_KEY };

async function convertFile(inputPath, outputFormat) {
  // Step 1: Submit conversion
  const formData = new FormData();
  formData.append("file", new Blob([fs.readFileSync(inputPath)]), path.basename(inputPath));
  formData.append("output_format", outputFormat);

  const response = await fetch(`${BASE_URL}/convert`, {
    method: "POST",
    headers,
    body: formData,
  });

  if (!response.ok) throw new Error(`Upload failed: ${response.status}`);
  const { job_id } = await response.json();
  console.log(`Job created: ${job_id}`);

  // Step 2: Poll for completion
  while (true) {
    const statusResp = await fetch(`${BASE_URL}/status/${job_id}`, { headers });
    const { status } = await statusResp.json();
    console.log(`Status: ${status}`);

    if (status === "completed") break;
    if (status === "failed") throw new Error("Conversion failed");
    await new Promise((r) => setTimeout(r, 2000));
  }

  // Step 3: Download
  const downloadResp = await fetch(`${BASE_URL}/download/${job_id}`, { headers });
  const buffer = Buffer.from(await downloadResp.arrayBuffer());
  const ext = path.extname(inputPath);
  const outputPath = inputPath.replace(ext, `.${outputFormat}`);
  fs.writeFileSync(outputPath, buffer);
  console.log(`Saved to ${outputPath}`);
  return outputPath;
}

async function useTool(inputPath, tool, params = {}) {
  const formData = new FormData();
  formData.append("file", new Blob([fs.readFileSync(inputPath)]), path.basename(inputPath));
  formData.append("tool", tool);
  for (const [key, value] of Object.entries(params)) {
    formData.append(key, value);
  }

  const response = await fetch(`${BASE_URL}/tool`, {
    method: "POST",
    headers,
    body: formData,
  });
  const { job_id } = await response.json();

  while (true) {
    const { status } = await (await fetch(`${BASE_URL}/status/${job_id}`, { headers })).json();
    if (status === "completed") break;
    if (status === "failed") throw new Error("Processing failed");
    await new Promise((r) => setTimeout(r, 2000));
  }

  const downloadResp = await fetch(`${BASE_URL}/download/${job_id}`, { headers });
  const buffer = Buffer.from(await downloadResp.arrayBuffer());
  const outputPath = `processed_${path.basename(inputPath)}`;
  fs.writeFileSync(outputPath, buffer);
  return outputPath;
}

async function checkUsage() {
  const resp = await fetch(`${BASE_URL}/usage`, { headers });
  const data = await resp.json();
  console.log(`Used: ${data.requests_today}/${data.rate_limit} today`);
  console.log(`Remaining: ${data.remaining}`);
}

module.exports = { convertFile, useTool, checkUsage };

// Usage:
// convertFile("image.png", "webp");
// useTool("document.pdf", "compress-pdf");
// useTool("photo.jpg", "resize-image", { width: "800", height: "600" });

File: examples/python/convert.py
"""
MegaConvert API — Python example using requests
Full docs: https://megaconvert.io/docs/api

Install: pip install requests
"""

import requests
import time

API_KEY = "mc_your_api_key_here"
BASE_URL = "https://megaconvert.io/api/v1"

headers = {"X-API-Key": API_KEY}


def convert_file(input_path: str, output_format: str) -> str:
    """Convert a file and return the output file path."""

    # Step 1: Submit conversion
    with open(input_path, "rb") as f:
        response = requests.post(
            f"{BASE_URL}/convert",
            headers=headers,
            files={"file": f},
            data={"output_format": output_format},
        )
    response.raise_for_status()
    job_id = response.json()["job_id"]
    print(f"Job created: {job_id}")

    # Step 2: Poll for completion
    while True:
        status_resp = requests.get(f"{BASE_URL}/status/{job_id}", headers=headers)
        status = status_resp.json()["status"]
        print(f"Status: {status}")

        if status == "completed":
            break
        elif status == "failed":
            raise Exception("Conversion failed")
        time.sleep(2)

    # Step 3: Download result
    download_resp = requests.get(f"{BASE_URL}/download/{job_id}", headers=headers)
    output_path = f"{input_path.rsplit('.', 1)[0]}.{output_format}"
    with open(output_path, "wb") as f:
        f.write(download_resp.content)

    print(f"Saved to {output_path}")
    return output_path


def use_tool(input_path: str, tool: str, **params) -> str:
    """Process a file with a tool (compress, resize, etc.)."""

    with open(input_path, "rb") as f:
        response = requests.post(
            f"{BASE_URL}/tool",
            headers=headers,
            files={"file": f},
            data={"tool": tool, **params},
        )
    response.raise_for_status()
    job_id = response.json()["job_id"]

    while True:
        status = requests.get(
            f"{BASE_URL}/status/{job_id}", headers=headers
        ).json()["status"]
        if status == "completed":
            break
        elif status == "failed":
            raise Exception("Processing failed")
        time.sleep(2)

    download_resp = requests.get(f"{BASE_URL}/download/{job_id}", headers=headers)
    output_path = f"processed_{input_path}"
    with open(output_path, "wb") as f:
        f.write(download_resp.content)

    return output_path


def check_usage():
    """Check your API usage."""
    resp = requests.get(f"{BASE_URL}/usage", headers=headers)
    data = resp.json()
    print(f"Used: {data['requests_today']}/{data['rate_limit']} today")
    print(f"Remaining: {data['remaining']}")


if __name__ == "__main__":
    # Convert PNG to WebP
    convert_file("image.png", "webp")

    # Compress a PDF
    # use_tool("document.pdf", "compress-pdf")

    # Resize an image
    # use_tool("photo.jpg", "resize-image", width="800", height="600")
